### PR TITLE
Only pull compile dependencies during compilation

### PR DESCRIPTION
**Pull request by alexarchambault**
Maven offers a way to pull some dependencies only at runtime, via [Maven scopes](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope).

Roughly speaking, in Maven, dependencies can be added
- only for compilation (`provided` scope)
- for compilation and runtime (`compile` scope - `compile` is for both compile-time and runtime)
- only at runtime (`runtime` scope)

This matters during dependency management: coursier can pull runtime dependencies or not, for example.

Mill already knows about compile-time versus runtime, with `compileIvyDeps` / `ivyDeps` / `runIvyDeps`, which are equivalent to Maven scopes `provided` / `compile` / `runtime`.

The PR here asks coursier not to pull runtime dependencies in Mill class paths made for compile-time, like `compileClasspath`.

### CODE REVIEW

**Comment by alexarchambault** [comment]
Wondering if this couldn't be handled better by introducing `CoursierModule#runResolutionParams` rather than passing a `coursier.core.Configuration` around…

**Comment by lihaoyi** [comment]
@alexarchambault could you elaborate a bit on the problem in the PR description and how you are solving it? Some of the code changes seem pretty non-obvious to me

File: kotlinlib/src/mill/kotlinlib/js/KotlinJsModule.scala [hunk 1]
```diff
@@ -210,7 +210,7 @@ trait KotlinJsModule extends KotlinModule { outer =>
       outputMode = binaryKindToOutputMode(kotlinJsBinaryKind()),
       irClasspath = Some(compile().classes),
       allKotlinSourceFiles = Seq.empty,
-      librariesClasspath = compileClasspath(),
+      librariesClasspath = upstreamAssemblyClasspath(),
```
Comment (lihaoyi) [inline kotlinlib/src/mill/kotlinlib/js/KotlinJsModule.scala:213]: e.g. what does this have to do with compile time deps?

Comment (alexarchambault) [reply kotlinlib/src/mill/kotlinlib/js/KotlinJsModule.scala:213]: Maybe we could have a new dedicated target, like `runDependencyClasspath` - with all dependency JARs, including JARs made for runtime, but excluding the current module classes. Or maybe we could just rename `upstreamAssemblyClasspath`, given it can have a broader uses.

**Comment by alexarchambault** [comment]
To expand the PR description, Android dependencies tend to pull more dependencies at runtime.

Compile-time:
```text
$ cs resolve androidx.compose.animation:animation-core:1.1.1 androidx.compose.ui:ui:1.1.1 -r google -c compile
androidx.annotation:annotation:1.2.0:compile
androidx.compose.animation:animation-core:1.1.1:compile
androidx.compose.runtime:runtime:1.1.1:compile
androidx.compose.runtime:runtime-saveable:1.1.1:compile
androidx.compose.ui:ui:1.1.1:compile
androidx.compose.ui:ui-geometry:1.1.1:compile
androidx.compose.ui:ui-graphics:1.1.1:compile
androidx.compose.ui:ui-text:1.1.1:compile
androidx.compose.ui:ui-unit:1.1.1:compile
org.jetbrains:annotations:13.0:compile
org.jetbrains.kotlin:kotlin-stdlib:1.5.30:compile
org.jetbrains.kotlin:kotlin-stdlib-common:1.5.30:compile
org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.5.30:compile
org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.5.30:compile
org.jetbrains.kotlinx:kotlinx-coroutines-android:1.5.2:compile
org.jetbrains.kotlinx:kotlinx-coroutines-core:1.5.2:compile
org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.5.2:compile
```

Runtime:
```text
$ cs resolve androidx.compose.animation:animation-core:1.1.1 androidx.compose.ui:ui:1.1.1 -r google
androidx.annotation:annotation:1.2.0:default
androidx.arch.core:core-common:2.1.0:default
androidx.arch.core:core-runtime:2.1.0:default
androidx.autofill:autofill:1.0.0:default
androidx.collection:collection:1.1.0:default
androidx.compose.animation:animation-core:1.1.1:default
androidx.compose.runtime:runtime:1.1.1:default
androidx.compose.runtime:runtime-saveable:1.1.1:default
androidx.compose.ui:ui:1.1.1:default
androidx.compose.ui:ui-geometry:1.1.1:default
androidx.compose.ui:ui-graphics:1.1.1:default
androidx.compose.ui:ui-text:1.1.1:default
androidx.compose.ui:ui-unit:1.1.1:default
androidx.compose.ui:ui-util:1.1.1:default
androidx.core:core:1.5.0:default
androidx.lifecycle:lifecycle-common:2.3.0:default
androidx.lifecycle:lifecycle-common-java8:2.3.0:default
androidx.lifecycle:lifecycle-runtime:2.3.0:default
androidx.lifecycle:lifecycle-viewmodel:2.3.0:default
androidx.profileinstaller:profileinstaller:1.1.0:default
androidx.savedstate:savedstate:1.1.0:default
androidx.startup:startup-runtime:1.0.0:default
androidx.tracing:tracing:1.0.0:default
androidx.versionedparcelable:versionedparcelable:1.1.1:default
org.jetbrains:annotations:13.0:default
org.jetbrains.kotlin:kotlin-stdlib:1.6.10:default
org.jetbrains.kotlin:kotlin-stdlib-common:1.6.10:default
org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.5.30:default
org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.5.30:default
org.jetbrains.kotlinx:kotlinx-coroutines-android:1.5.2:default
org.jetbrains.kotlinx:kotlinx-coroutines-core:1.5.2:default
org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.5.2:default
```
